lib/AST/ASTImporter.cpp

```diff
@@ -247,42 +254,47 @@ namespace clang {
-    bool ImportDefinition(ObjCInterfaceDecl *From, ObjCInterfaceDecl *To,
-                          ImportDefinitionKind Kind = IDK_Default);
-    bool ImportDefinition(ObjCProtocolDecl *From, ObjCProtocolDecl *To,
-                          ImportDefinitionKind Kind = IDK_Default);
-    TemplateParameterList *ImportTemplateParameterList(
+    Error ImportDefinition(
+        RecordDecl *From, RecordDecl *To,
+        ImportDefinitionKind Kind = IDK_Default);
+    Error ImportDefinition(
+        VarDecl *From, VarDecl *To,
+        ImportDefinitionKind Kind = IDK_Default);
+    Error ImportDefinition(
+        EnumDecl *From, EnumDecl *To,
+        ImportDefinitionKind Kind = IDK_Default);
+    Error ImportDefinition(
+        ObjCInterfaceDecl *From, ObjCInterfaceDecl *To,
+        ImportDefinitionKind Kind = IDK_Default);
+    Error ImportDefinition(
+        ObjCProtocolDecl *From, ObjCProtocolDecl *To,
+        ImportDefinitionKind Kind = IDK_Default);
+    Expected<TemplateParameterList *> ImportTemplateParameterList(
         TemplateParameterList *Params);
-    TemplateArgument ImportTemplateArgument(const TemplateArgument &From);
-    Optional<TemplateArgumentLoc> ImportTemplateArgumentLoc(
+    Expected<TemplateArgument> ImportTemplateArgument(
+        const TemplateArgument &From);
+    Expected<TemplateArgumentLoc> ImportTemplateArgumentLoc(
         const TemplateArgumentLoc &TALoc);
-    bool ImportTemplateArguments(const TemplateArgument *FromArgs,
-                                 unsigned NumFromArgs,
-                                 SmallVectorImpl<TemplateArgument> &ToArgs);
+    Error ImportTemplateArguments(
+        const TemplateArgument *FromArgs, unsigned NumFromArgs,
+        SmallVectorImpl<TemplateArgument> &ToArgs);
 
     template <typename InContainerTy>
-    bool ImportTemplateArgumentListInfo(const InContainerTy &Container,
-                                        TemplateArgumentListInfo &ToTAInfo);
+    Error ImportTemplateArgumentListInfo(
+        const InContainerTy &Container, TemplateArgumentListInfo &ToTAInfo);
 
     template<typename InContainerTy>
-    bool ImportTemplateArgumentListInfo(SourceLocation FromLAngleLoc,
-                                        SourceLocation FromRAngleLoc,
-                                        const InContainerTy &Container,
-                                        TemplateArgumentListInfo &Result);
+    Error ImportTemplateArgumentListInfo(
+      SourceLocation FromLAngleLoc, SourceLocation FromRAngleLoc,
+      const InContainerTy &Container, TemplateArgumentListInfo &Result);
 
     using TemplateArgsTy = SmallVector<TemplateArgument, 8>;
-    using OptionalTemplateArgsTy = Optional<TemplateArgsTy>;
-    std::tuple<FunctionTemplateDecl *, OptionalTemplateArgsTy>
-    ImportFunctionTemplateWithTemplateArgsFromSpecialization(
-        FunctionDecl *FromFD);
+    Error ImportFunctionTemplateWithTemplateArgsFromSpecialization(
```

Comment:
`Expected<std::tuple<FunctionTemplateDecl *, OptionalTemplateArgsTy>>` ?

Reply:
A `FunctionTemplateDecl *` and a `TemplateArgsTy` are two output values from the function. But if any of these fails to import the whole function will fail (or not?), there is no valid case when there was no import error and one of the output arguments is null (empty optional or null pointer). So optional is not needed at return, and it is more simple to have output parameters like at every other similar function.
There are two kind of import function, one that has a reference output value and returns `Error` and one that returns `Expected`. The first should update something in an existing object, the second type should return something new. (For this reason a tuple is the better solution here, only it was somewhat inconvenient.)

Reply:
Yeah I think we agree. My understanding is that we should return `Error` in case of functions which previously did not return with any meaningful value, e.g. `void` or `bool` to indicate error. In those cases where the function returns with a  meaningful value we should use `Expected`.

Reply:
Still it is hard to decide when to use the `Error` return value and when to use `Expected`. If a function has a vector output value it may be not efficient (?) to return it with Expected. For `SmallVector` the used "size" of the vector is different when `ImportFunctionTemplateWithTemplateArgsFromSpecialization` is called, the parameter vector is not size-dependent but a returned vector has the same size ("size": template parameter for `SmallVector`). Probably it is better to make everything work with output parameters and `Error` return value.

Reply:
>  If a function has a vector output value it may be not efficient (?) to return it with Expected.

I think, if the cost of returning T is 1 unit than the cost of returning Expected<T> will be 1 unit too. I mean it should be if the template is written properly.

> For SmallVector the used "size" of the vector is different when ImportFunctionTemplateWithTemplateArgsFromSpecialization is called, the parameter vector is not size-dependent but a returned vector has the same size ("size": template parameter for SmallVector). 

I don't quite understand this. The size template parameter of `SmallVector` should be the same, namely `OptionalTemplateArgsTy`.

> Probably it is better to make everything work with output parameters and Error return value.

I don't think so, I rather think the receipt is very simple: if T is not `void` or `bool` then `T f() ---> Expected<T> f()` , otherwise use `Error`.

Reply:
The problem is with `ImportTemplateArguments`, it has currently a vector reference output parameter but it should return this value with `Expected`. This function is used with `TemplateArgsTy` and with other `SmallVector` (with size 2).

Reply:
And the `ImportDeclParts` function should be changed to return a big tuple inside an `Expected`?

Reply:
TLDR, we can have the output params, it is OK.


Just some other thoughts:

> And the ImportDeclParts function should be changed to return a big tuple inside an Expected?

Okay, `ImportDeclParts` is already taking output parameters.
But for me it would be easier to understand the code if we never used output parameters, especially not with reference types. Because it is very hard to see on the callee site that we pass an output parameter. That is why I used previously tuples to return multiple values.

So, I'd rather have `ImportDeclParts` returning with a tuple. But then we should change also `GetImportedOrCreateDecl` and who knows what else.

Reply:
> The problem is with ImportTemplateArguments, it has currently a vector reference output parameter but it should return this value with Expected. This function is used with TemplateArgsTy and with other SmallVector (with size 2).

We can use `SmallVectorImpl` in such cases:
In APIs that don’t care about the “small size” (most?), prefer to use the SmallVectorImpl<T> class, which is basically just the “vector header” (and methods) without the elements allocated after it. Note that SmallVector<T, N> inherits from SmallVectorImpl<T> so the conversion is implicit and costs nothing.

Reply:
I do not want refactoring now, just "mechanical" changes. The `ImportFunctionTemplateWithTemplateArgsFromSpecialization` will return tuple in Expected, and what returned `void` or `bool` will return `Error`.

Reply:
If we have an `Expected<T>` as a result type, then there is no way to acquire the `T` if there was an error. However, when we have output parameters and `Error`, then even if there was an error some output params could have been set ... and those can be reused even after the return with the `Error`. If error handling is not proper on the callee site then we may continue with stale values, which is not possible if we use `Expected<T>`.

And I think the above is quite a strong argument against output params. Now I changed my mind, could we just remove all output params?

Reply:
> And I think the above is quite a strong argument against output params. Now I changed my mind, could we just remove all output params?

Or at least maybe we should just not add new functions which uses output params. (Maybe later we could change existing functions)

Reply:
> I do not want refactoring now, just "mechanical" changes. The ImportFunctionTemplateWithTemplateArgsFromSpecialization will return tuple in Expected, and what returned void or bool will return Error.

Ok, that's fine for me.

Reply:
Another thing to refactor: Make the input parameters of import functions `const`.

Reply:
:+1: